src/server/game/AI/SmartScripts/SmartScript.cpp

```diff
@@ -1123,22 +1123,39 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
             delete targets;
             break;
         }
-        case SMART_ACTION_MOVE_FORWARD:
+        case SMART_ACTION_MOVE_OFFSET:
         {
-            if (!me)
-                break;
+            if (ObjectList* targets = GetTargets(e, unit))
+            {
+                for (ObjectList::const_iterator itr = targets->begin(); itr != targets->end(); ++itr)
+                {
+                    if (!IsCreature(*itr))
+                        continue;
 
-            float x, y, z;
-            me->GetClosePoint(x, y, z, me->GetObjectSize() / 3, (float)e.action.moveRandom.distance);
-            me->GetMotionMaster()->MovePoint(SMART_RANDOM_POINT, x, y, z);
-            break;
-        }
-        case SMART_ACTION_RISE_UP:
-        {
-            if (!me)
-                break;
+                    Position pos = (*itr)->GetPosition();
+
+                    if (e.action.offsetMovement.movement)
+                    {
+                        // Use normal WoW axis-based movement
+                        Position offset = { e.target.x, e.target.y, e.target.z, 0.0f };
+                        pos.RelocateOffset(offset);
+                        (*itr)->ToCreature()->GetMotionMaster()->MovePoint(SMART_RANDOM_POINT, pos);
+                    }
+                    else
+                    {
+                        // Use forward/backward/left/right cartesian plane movement
+                        float x, y, z, o;
+                        o = pos.GetOrientation();
+                        x = pos.GetPositionX() + (cos(o - (M_PI / 2))*e.target.x) + (cosf(o)*e.target.y);
+                        y = pos.GetPositionY() + (sin(o - (M_PI / 2))*e.target.x) + (sinf(o)*e.target.y);
```

Comment:
You mean this calculation? https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Entities/Vehicle/VehicleDefines.h#L132